### DISCUSSION TITLE
$subscribed not triggered when an optional property is set first time

### DISCUSSION THREAD

**Discussion by ygyg70** [vuejs/pinia#1192] [Help and Questions]
### Reproduction

-

### Steps to reproduce the bug

1) Create a store with an optional property, such as:

export interface StateInterface {
  prop?:string
}
export const useStore = defineStore( 'store', {
  state: ():StateInterface => ( { } )
})

2) Subscribe to the store
3) Set the value of store.prop, subscribe will not trigger.

Note - if $patch is used to set the value, subscribe will trigger

### Expected behavior

$subscribe will trigger when a value is set for the first time

### Actual behavior

For $subscribe to trigger, property has to be initiated as null

**Comment by dreambo8563** [comment]
@ygyg70 provide additional info:
if set the prop with `u.$state.prop = "100"; . that will trigger subscribe which actually invoke the $patch

set value to optional prop won't trigger the watch , I don't know how to make it work.   @posva , could you pls help on this? It seems lots of question on optional property.
 https://github.com/vuejs/pinia/blob/2b998ee25832ccebf132c677fe288cc9a775de05/packages/pinia/src/store.ts#L391

**Reply by posva** [reply]
This is expected. It’s good practice to declare initial state

**Comment by ygyg70** [comment]
I switched from Vuex to Pinia, had to spend time to troubleshoot.
Expect others to encounter the same, so at least should be included in the migration docs

**Reply by dreambo8563** [reply]
I read vuex doc again.

$state.xx =xx will do the replace. it's better to have initial state

>Prefer initializing your store's initial state with all desired fields upfront.
When adding new properties to an Object, you should either:
Use Vue.set(obj, 'newProp', 123), or
- Replace that Object with a fresh one. For example, using the [object spread syntax](https://github.com/tc39/proposal-object-rest-spread) we can write it like this:
